### PR TITLE
first fixes for #1348

### PR DESCRIPTION
**Pull request by rogergl**
Some problems I have encountered so far are:
1. I have no idea how to run the tests
2. The formatting is different
3. nrepl-client requires cider-util and cider-util requires nrepl-client

### CODE REVIEW

**Comment by bbatsov** [comment]
> I have no idea how to run the tests

make test

> The formatting is different

Well, that's normal. Changes require formatting changes.

> 3. nrepl-client requires cider-util and cider-util requires nrepl-client

Not sure how this is relevant.

File: cider-browse-ns.el [hunk 1]
```diff
@@ -131,13 +132,13 @@ contents of the buffer are not reset before inserting TITLE and ITEMS."
 (defun cider-browse-ns--doc-at-point ()
   "Expand browser according to thing at current point."
   (interactive)
-  (-when-let (var (cider-browse-ns--var-at-point))
-    (cider-doc-lookup var)))
+  (when-let (var (cider-browse-ns--var-at-point))
+            (cider-doc-lookup var)))
```
Comment (bbatsov) [inline cider-browse-ns.el]: This looks like some bug.

**Comment by expez** [comment]
> Well, that's normal. Changes require formatting changes.

This is not OK imo.  It looks incredibly bad when `& body` isn't indented like usual.  Why don't these have indention specs?

File: cider-browse-ns.el (hunk 1 above)
Comment (rogergl) [inline cider-browse-ns.el]: What exactly ?

**Comment by expez** [comment]
hmm they do have indentation specs.  @rogergl you need to load the macros and re-indent these files so the indentation spec is used.

**Comment by rogergl** [comment]
@expez I'll must admit that I have no idea what indentation specs are and which macros to load. Is there some guide that describes it  ?

**Comment by rogergl** [comment]
@bbatsov make test (running on Mac OS X) gives
```
GNU Emacs 24.5.1
Copyright (C) 2015 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
cask install
make: cask: Permission denied
make: *** [elpa] Error 1
```

**Comment by expez** [comment]
> I'll must admit that I have no idea what indentation specs are. 

Below, after the docstring, is the indent spec. 
```el   
 (defmacro when-let (bindings &rest body)
      "Process BINDINGS and if all values are non-nil eval BODY.
Argument BINDINGS is a list of tuples whose car is a symbol to be
bound and (optionally) used in BODY, and its cadr is a sexp to be
evalled to set symbol's value.  In the special case you only want
to bind a single value, BINDINGS can just be a plain tuple."
      (declare (indent 1) (debug if-let))
      (list 'if-let bindings (macroexp-progn body))))
```

> [...]  and which macros to load

You must load macros you're substituting in, like `when-let` above.  Otherwise emacs won't know how to indent these forms.

**Comment by rogergl** [comment]
The formatting is fixed. Thanks @expez

**Comment by bbatsov** [comment]
@rogergl Did you install `cask`?

**Comment by rogergl** [comment]
@bbatsov "cider-util should not have anything nrepl-related in it.: should I remove it ?

**Comment by bbatsov** [comment]
> should I remove it ?

Yep.

**Comment by Malabarba** [comment]
I still need to go through the whole diff, but I just wanted to say that we need to define these new macros on their own file, otherwise we might run into that problem of faulty installations again.

**Comment by bbatsov** [comment]
We can always have `cider-compat.el` or something like this.

**Comment by Malabarba** [comment]
Another thing. Even if there's just one binding, I much prefer we use `(if-let ((a b)) ...)` instead of `(if-let (a b) ...)`, for general consistency with `let` and with the many-bindings case. Dash's `-if-let` didn't allow this, but `if-let` does, so let's do it. :+1: 
Of course, the same goes for `when-let`.

**Comment by Malabarba** [comment]
Also, if I haven't said it yet. Thanks for all this work @rogergl .

**Comment by Malabarba** [comment]
> nrepl-client requires cider-util and cider-util requires nrepl-client

This might fix itself if we move the macros to cider-compat. Still, Bozhidar is right that anything nrepl dependent should be on cider-client, not cider-util.

**Comment by rogergl** [comment]
Current to-dos: use (if-let ((a b)) ...) etc
Move the macros to cider-comapt.el.  Should I make this changes ?

**Comment by Malabarba** [comment]
Yep. 👍

**Comment by bbatsov** [comment]
It seems we're almost ready here. There are only some byte-compilation warnings that have to be fixed. 

It would be nice to wrap the PR as soon as possible, because it touches a lot of places in the code and any potential change right now would result in conflicts for the PR.

**Comment by rogergl** [comment]
I could spend some hours tomorrow. cider-compat or cider-backports ? Which one is preferred ?

**Comment by bbatsov** [comment]
Let's go with `cider-compat`.

**Comment by rogergl** [comment]
Ok, the byte compilation error are fixed. I think the change to use (if-let ((a b)) ...) instead of (if-let (a b) ...) could be done in another migration ? Leaving just the following remaining changes:

cider-eldoc using -elem-index
cider-interaction using -each
cider-stacktrace using -repeat
cider-util -using map-indexed

**Comment by bbatsov** [comment]
Great news! Well, I guess you can just squash the commits together when you feel you're ready. Use the project convention for commit message structuring for the final commit message.

**Comment by rogergl** [comment]
Use the project convention for commit message structuring for the final commit message.
I think I need a little help with this, is there a guide that describes the required steps ?

What about 

    cider-eldoc using -elem-index
    cider-interaction using -each
    cider-stacktrace using -repeat
    cider-util -using map-indexed

Next iteration or moving them from dash to compat ?

**Comment by rogergl** [comment]
Ok just found the contributing readme.

**Comment by bbatsov** [comment]
> Next iteration or moving them from dash to compat ?

`-each` is basically `mapc`, so you've got that one covered. You can moved the rest to `cider-util` and rename them to `cider-elem-index`, etc.

**Comment by rogergl** [comment]
Since there is already some stuff from dash in cider-compat, shouldn't they be moved there ?

**Comment by bbatsov** [comment]
> Since there is already some stuff from dash in cider-compat, shouldn't they be moved there ?

Nope. `compat` is for compatibility with older versions of Emacs (it should have only things backported from Emacs itself). `util` is for general purpose functions that we use all over cider.

**Comment by rogergl** [comment]
But none of the thing in emacs-compat is backported. The function there are  just copies from dash and subr-x ?

**Comment by bbatsov** [comment]
If you copy from from `subr-x`, it's considered backported, as this is a built-in library. Things from `dash` should be go to util and be prefixed with `cider-`, as it's a 3rd party library.

**Comment by bbatsov** [comment]
Btw, it seems to me everything in `compat` right now is from `subr-x`, so we're great.

**Comment by Malabarba** [comment]
It looks like `cl-position` is a drop-in replacement for `-elem-index`.
Since we depend on emacs-24.3 we don't even have to depen on cl-lib to use it.

**Comment by rogergl** [comment]
Dash is gone. Leaving 2 issues:

1. ) I had to copy the --each macro. I tried to convert cider--map-indexed to use maps but wasn't successful.

2.) I realy could need some help on how to turn all this commits into a valid pull request.

**Comment by bbatsov** [comment]
> I had to copy the --each macro. I tried to convert cider--map-indexed to use mapc but wasn't successful.

We use it very rarely, so just expand it to not use the magic `it` variable. That's the only difference between a regular `-each` and `--each`. 

See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

After you make the changes locally use `git push -f` to update your PR.

**Comment by rogergl** [comment]
We use it very rarely, so just expand it to not use the magic it variable. 
That goes probally beyond the things I currently understand about Emacs macro programing. But I'll have a look at it tomorrow.

**Comment by Malabarba** [comment]
In a couple of hours I can give you a standalone implementation of -map-indexed, using cl-mapcar.

**Comment by rogergl** [comment]
BTW:  cider-map-index is only used by cider-intern-keys with is only used by cider-test

**Comment by Malabarba** [comment]
(defun cider-map-indexed (f list)
      "Return a list of (F index item) for each item in LIST."
      (let ((i 0)
            (out))
        (dolist (it list (nreverse out))
          (push (funcall f i it) out)
          (setq i (1+ i)))))

**Comment by rogergl** [comment]
I think cider-repeat can be replaced with make-list ?

**Comment by Malabarba** [comment]
Yes. And that should allow you to get rid of `!cons` and `cider--dotimes` too.

**Comment by rogergl** [comment]
I made a new pull request. One thing that needs fixing is probably the copyright notice in cider-combat...

File: CHANGELOG.md [hunk 2]
```diff
@@ -38,6 +38,8 @@
 
 ### Changes
 
+* [#1348](https://github.com/clojure-emacs/cider/issues/1348) dash is no longer neede.
+
```
Comment (bbatsov) [inline CHANGELOG.md]: I'd refresh this to "Drop the `dash` dependency.". Also - kill the blank line.

**Comment by bbatsov** [comment]
In the commit message - remove -> Remove

You'll also have to rebase this branch of top of the current `master`.

**Comment by rogergl** [comment]
You'll also have to rebase this branch of top of the current master
How does that work ? Just call git rebase master ?

**Comment by bbatsov** [comment]
More or less. You'll have to update your local `master` first.

**Comment by bbatsov** [comment]
Remove dash dependencies -> Remove dash dependency

You'll also have to squash the commits once again, but we're almost ready.

File: cider-mode.el [hunk 3]
```diff
@@ -263,7 +264,7 @@ Returns to the buffer in which the command was invoked."
         ("Find"
          ["Find definition" cider-find-var]
          ["Find resource" cider-find-resource]
-         ["Go back" cider-pop-back])
+         ["Jump back" cider-jump-back])
```
Comment (bbatsov) [inline cider-mode.el]: Seems you resolved to conflict here incorrectly.

File: cider-mode.el [hunk 4]
```diff
@@ -275,7 +276,6 @@ Returns to the buffer in which the command was invoked."
         ["Toggle var tracing" cider-toggle-trace-var]
         ["Toggle ns tracing" cider-toggle-trace-ns]
         ["Refresh loaded code" cider-refresh]
-        ["Select any CIDER buffer" cider-selector]
```
Comment (bbatsov) [inline cider-mode.el]: Here as well.

File: cider-mode.el [hunk 5]
```diff
@@ -297,24 +297,22 @@ Returns to the buffer in which the command was invoked."
         ["Quit" cider-quit]
         ["Restart" cider-restart]
         "--"
-        ["View manual online" cider-open-manual]
```
Comment (bbatsov) [inline cider-mode.el]: Here as well.

File: cider-mode.el [hunk 6]
```diff
@@ -297,24 +297,22 @@ Returns to the buffer in which the command was invoked."
...
+  (when-let (indent
+             (nrepl-dict-get (cider-resolve-var (cider-current-ns) symbol-name)
+                             "indent"))
     (let ((format (format ":indent metadata on ‘%s’ is unreadable! \nERROR: %%s"
                           symbol-name)))
       (with-demoted-errors format
         (cider--deep-vector-to-list (read indent))))))
 
 ;;; Dynamic font locking
-(defcustom cider-font-lock-dynamically '(macro core deprecated)
```
Comment (bbatsov) [inline cider-mode.el]: Here as well.

**Comment by bbatsov** [comment]
You deleted some new code by mistake, while you were resolving your merge conflicts. Please go over your diff one more time and fix this.

**Comment by rogergl** [comment]
I used Sourcetree and said keep their changes. Don't know what went wrong. I try to fix this.

**Comment by Malabarba** [comment]
Don't give up! And remember, you're not wasting time, you're learning git. ;-)

**Comment by rogergl** [comment]
I accidentally committed an additional file now I have to figure out how to undo this.

**Comment by rogergl** [comment]
I tried again but probably messed up the commit history since I accidentally committed a cider-mode-old.el file.

**Comment by bbatsov** [comment]
The code looks fine to me right now. @Malabarba can you go over it as well before we merge it?

**Comment by Malabarba** [comment]
Ok, I see only one issue. Inside `cider-tests--no-auto.el` you require `cider-contrib` instead of `cider-compat`.

**Comment by Malabarba** [comment]
I see many small things I'd like to fix, but they are mostly orthogonal to this PR (they were already there before) so I can do them myself after we merge this.

**Comment by bbatsov** [comment]
I'll merge this and force-push to fix the contrib mistake.

**Comment by Malabarba** [comment]
Thanks again @rogergl

**Comment by rogergl** [comment]
@Malabarba: Thanks for given me the chance to learn more about git ;-)

**Comment by bbatsov** [comment]
:-)

**Pull request merged**